llms/mlx_lm/utils.py

```diff
@@ -85,10 +83,34 @@ def get_model_path(path_or_hf_repo: str) -> Path:
     return model_path
 
 
+def apply_penalty(logits: mx.array, generated_tokens: Any, penalty: float):
+    """
+    Apply repetition penalty to specific logits based on the given context.
+
+    Args:
+        logits (mx.array): The logits produced by the language model.
+        generated_tokens (any): A list of N previous tokens.
+        penalty (float): The repetition penalty factor to be applied.
+
+    Returns:
+        None: The function modifies the input logits in place.
+    """
+    if len(generated_tokens) > 0:
+        indices = mx.array([token for token in generated_tokens])
+        selected_logits = logits[:, indices]
+        selected_logits = mx.where(
+            selected_logits < 0, selected_logits * penalty, selected_logits / penalty
+        )
+        logits[:, indices] = selected_logits
+    return logits
+
+
 def generate_step(
     prompt: mx.array,
     model: nn.Module,
-    temp: float = 0.0,
+    temp: 0.0,
+    repetition_penalty: Optional[float],
```

Comment:
Issue: Assign default values for `repetition_penalty` and `repetition_context` as well, in case some users directly imports generate_step and uses it, so that we don't break their code.